target/linux/ath79/image/generic.mk

```diff
@@ -314,6 +342,28 @@ define Device/avm_fritzdvbc
 endef
 TARGET_DEVICES += avm_fritzdvbc
 
+define Device/belkin_f9j1108-v2
+  SOC := qca9558
+  DEVICE_VENDOR := Belkin
+  DEVICE_MODEL := AC1750 DB Wi-Fi (F9J1108v2)
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
+  KERNEL_SIZE := 5120k
+  LOADER_TYPE := bin
+  LOADER_FLASH_OFFS := 0x50000
+  COMPILE := loader-$(1).bin loader-$(1).uImage
+  COMPILE/loader-$(1).bin := loader-okli-compile
+  COMPILE/loader-$(1).uImage := append-loader-okli $(1) | pad-to 64k | lzma | \
+	uImage lzma
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma -M 0x4f4b4c49
+  IMAGE/sysupgrade.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-rootfs | \
+	pad-rootfs | check-size
+  IMAGE/factory.bin := create-edimax-factory-bin F9J1108v1 BR-6679BAC \
+    $(KDIR)/loader-$(1).uImage $(KDIR_TMP)/openwrt-ath79-generic-$(1)-squashfs-sysupgrade.bin
```

Comment:
I wonder whether you couldn't reuse more of the existing statements here instead of reinventing the wheel and putting everything in this command.

Reply:
I was attempting to write it in such a way that would be reusable for other Edimax devices, any specific hints around how to achieve that reuse ?

Reply:
I think each step of your factory.bin build recipe should be a separate recipe. Everything except the headers can probably be replaced with append-kernel, append-rootfs and check-size. Everything before that can change and replace the images so that the generic commands uses the one with headers

Reply:
`$(KDIR)/loader-$(1).uImage $(KDIR_TMP)/openwrt-ath79-generic-$(1)-squashfs-sysupgrade.bin`

there is nothing in these parameters that is unique to any device, so it can be in the build recipe, in other words these parameters are not useful at all

Reply:
What value is there in splitting up the factory.bin recipe?  Are you saying it should be implemented like "append-kernel | convert_to_edimax uImage F9J1108v1 BR-6679BAC | append-rootfs | convert_to_edimax rootfs F9J1108v1 BR-6679BAC | check-size" ?  To me, that seems to violate what "append" means, since you are not appending the kernel to the output image, you need to modify it prior to appending.

The loader-$(1).uImage may not be named the same in other recipes, and some devices might want a different image passed in as the rootfs, so I made these parameters to cope with this.

Ie, if the other device uses a standard "denx,uimage" firmware partition, then both of these file names will be different.

Reply:
It would be the reverse. first the recipes that add the headers and replace the image files, then the append commands

Reply:
OK, so you are suggesting to re-implement edimax_fw_header.c, which includes header and data checksums, in a Makefile ?  Do we have any examples of checksum calculation for other devices ?

Reply:
using what you have this should work but I can't test it

```
define Build/edimax-headers
	$(eval edimax_magic=$(word 1,$(1)))
	$(eval edimax_model=$(word 2,$(1)))

	$(STAGING_DIR_HOST)/bin/edimax_fw_header -M $(edimax_magic) -m $(edimax_model)\
		-v $(VERSION_DIST)$(firstword $(subst +, , $(firstword $(subst -, ,$(REVISION))))) \
		-n "uImage" \
		-i $(KDIR)/loader-$(1).uImage \
		-o $$$$(IMAGE_KERNEL);
	$(STAGING_DIR_HOST)/bin/edimax_fw_header -M $(edimax_magic) -m $(edimax_model)\
		-v $(VERSION_DIST)$(firstword $(subst +, , $(firstword $(subst -, ,$(REVISION))))) \
		-n "rootfs" \
		-i $(KDIR_TMP)/openwrt-ath79-generic-$(1)-squashfs-sysupgrade.bin \
		-o $$$$(IMAGE_ROOTFS);
```

Reply:
OK, thanks for that, taking what you've done, I've reworked it so that it can install cleanly, could you please check ?